src/query/service/src/interpreters/interpreter_copy_v2.rs

```diff
@@ -164,40 +164,70 @@ impl CopyInterpreterV2 {
         tbl_name: &String,
         from: &ReadDataSourcePlan,
         files: Vec<String>,
-    ) -> Result<Vec<DataBlock>> {
-        let ctx = self.ctx.clone();
-        let settings = self.ctx.get_settings();
+    ) -> Result<PipelineBuildResult> {
+        let mut build_res = PipelineBuildResult::create();
 
-        let mut pipeline = Pipeline::create();
-        let read_source_plan = from.clone();
-        let read_source_plan = Self::rewrite_read_plan_file_name(read_source_plan, files);
+        let read_source_plan = Self::rewrite_read_plan_file_name(from.clone(), &files);
         tracing::info!("copy_files_to_table from source: {:?}", read_source_plan);
-        let table = ctx.build_table_from_source_plan(&read_source_plan)?;
-        let res = table.read2(ctx.clone(), &read_source_plan, &mut pipeline);
-        if let Err(e) = res {
-            return Err(e);
-        }
 
-        let table = ctx.get_table(catalog_name, db_name, tbl_name).await?;
+        let from_table = self.ctx.build_table_from_source_plan(&read_source_plan)?;
+        from_table.read2(
+            self.ctx.clone(),
+            &read_source_plan,
+            &mut build_res.main_pipeline,
+        )?;
+
+        let to_table = self.ctx.get_table(catalog_name, db_name, tbl_name).await?;
+
+        to_table.append2(self.ctx.clone(), &mut build_res.main_pipeline, false)?;
 
-        table.append2(ctx.clone(), &mut pipeline, false)?;
-        pipeline.set_max_threads(settings.get_max_threads()? as usize);
+        let ctx = self.ctx.clone();
+        let catalog_name = catalog_name.clone();
+        let files = files.clone();
+        let from = from.clone();
+
+        build_res.main_pipeline.set_on_finished(move |may_error| {
+            if may_error.is_none() {
+                // capture out variable
+                let ctx = ctx.clone();
+                let files = files.clone();
+                let from = from.clone();
+                let catalog_name = catalog_name.clone();
+                let to_table = to_table.clone();
+
+                let task = GlobalIORuntime::instance().spawn(async move {
+                    // Commit
+                    let operations = ctx.consume_precommit_blocks();
+                    to_table
+                        .commit_insertion(ctx.clone(), &catalog_name, operations, false)
+                        .await?;
+
+                    // Purge
+                    CopyInterpreterV2::purge_files(ctx, &from, &files).await
+                });
+
+                return match futures::executor::block_on(task) {
```

Comment:
Is blocking on current thread ok?

Reply:
It's ok. we blocking and wait for commit.